### ISSUE TITLE
Doing `.ckLocal()` on Singleton proxies sometimes returns nullptr

### ISSUE THREAD

**Issue by knelli2** [sxs-collaboration/spectre#3933]
# Bug reports:

### Expected behavior:

On a singleton component, I can call
```
auto& proxy = Parallel::get_parallel_component<SingletonComponent>(cache);
auto* local_ptr = proxy.ckLocal();
```
and have `local_ptr` point to the local object.

### Current behavior:

Sometimes, but not all the time, the pointer returned by `.ckLocal()` from a singleton is a `nullptr`, even when I'm running on the core that the singleton is on.

### Environment:

Using clang-10, tested with both charm v6 and charm v7. Both exhibit this issue.

### Detailed discussion:

https://github.com/knelli2/spectre/tree/cklocal_sing_error

To reproduce, this is a branch (based on develop) with minimal changes to the `InitializeInterpolationTarget` action. I picked an action I knew for certain was run on a singleton so that there wouldn't be any doubt that `.ckLocal()` should return a valid pointer. This is the only addition I made to the initialization action:
```
Parallel::printf("Initializing singleton %s on proc %d\n",
                        pretty_type::name<ParallelComponent>(), sys::my_proc());
auto& proxy = Parallel::get_parallel_component<ParallelComponent>(cache);
auto* ptr = proxy.ckLocal();
if (ptr == nullptr) {
  ERROR("Singleton " << pretty_type::name<ParallelComponent>(
                                   << " pointer in Initialization is null.");
}
```

Build and run the `EvolveGhKerrSchild` executable with the default input file `tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml`. This only happens when running on multiple cores (because on 1 core everything is local), and it doesn't happen all the time which leads me to believe this is some issue with charm singletons. When running on 2 cores, I hit the ERROR line above once out of 500 tries. On 4 cores, I hit the ERROR line above 58 out of 100 times.

I believe #3787 will fix this issue as we won't have any more charm singletons in our code.

Also, I think the reason this hasn't been detected so far is because we hardly ever get the local branch of a singleton. We almost exclusively call `.ckLocal()` on the DgElementArray.

I have also seen this once before when doing BBH testing with the `ControlComponent` singleton for control systems. Definitely running on the ControlComponent, I'd get a pointer to the local object and it would be a nullptr.

**Comment by nilsdeppe** [comment]
I don't see how you could have a race condition on a singleton. Race conditions happen when you have multiple threads accessing the same object. Singletons are tied to a specific thread.

**Comment by knelli2** [comment]
OK yeah maybe race condition wasn't the right thing to call this. I've edited the title and description.

**Comment by knelli2** [comment]
This issue is fixed by #3787.

I did another 500 runs of the KerrSchild executable on 4 cores and didn't see this issue once, so I'm fairly confident this is now fixed.
